astropy/io/misc/pyarrow/csv.py

```diff
@@ -0,0 +1,377 @@
+        there is no header line and the column names are taken from ``names`` or
+        generated automatically ("f0", "f1", ...).
+    data_start : int, None, optional (default None)
+        Line index for the start of data. If `None`, then data starts one line after the
+        header line, or on the first line if there is no header.
+    names : list, None, optional (default None)
+        List of names for input data columns when there is no header line. If supplied,
+        then ``header_start`` must be `None`.
+    include_names : list, None, optional (default None)
+        List of column names to include in output. If `None`, all columns are included.
+    dtypes : dict, None, optional (default None)
+        If provided, this is a dictionary of data types for output columns. Each key is
+        a column name and the value is a data type object that is accepted as an
+        argument to `np.dtype`. Examples include ``int``, ``np.float32``,
+        ``np.dtype('f4')`` or ``"float32"``. Default is to infer the data types.
+    comment: 1-character str or None, optional (default None)
+        Character used to indicate the start of a comment. Any line starting with
+        optional whitespace and then this character is ignored. Using this option will
+        cause the parser to be slower and use more memory as it uses Python code to
+        strip comments.
+
+    Other Parameters
+    ----------------
+    null_values : list, optional (default None)
+        List of strings to interpret as null values. By default, only empty strings are
+        considered as null values.
+    encoding: str, optional (default 'utf-8')
+        Encoding of the input data.
+    newlines_in_values: bool, optional (default False)
+        Whether newline characters are allowed in CSV values. Setting this to True
+        reduces the performance of multi-threaded CSV reading.
+
+    Returns
+    -------
+    Table
+        An astropy Table containing the data from the CSV file.
+    """
+    pa, csv = get_pyarrow_csv()
+
+    parse_options = pa.csv.ParseOptions(
+        delimiter=delimiter,
+        quote_char=quotechar,
+        double_quote=double_quote,
+        escape_char=escape_char,
+        newlines_in_values=newlines_in_values,
+    )
+
+    read_options = get_read_options(header_start, data_start, names, encoding)
+    convert_options = get_convert_options(include_names, dtypes, null_values)
+
+    if comment is not None:
+        input_file = strip_comment_lines(input_file, comment, encoding)
+
+    table_pa = csv.read_csv(
+        input_file,
+        parse_options=parse_options,
+        read_options=read_options,
+        convert_options=convert_options,
+    )
+    table_apt = convert_pa_table_to_astropy_table(table_pa)
+
+    return table_apt
+
+
+def get_pyarrow_csv():
+    """Helper function to import pyarrow and pyarrow.csv."""
+    from astropy.utils.compat.optional_deps import HAS_PYARROW
```

Comment:
I think this can be on top - that does not actually import but just checks whether something is present.

I'd probably move the function up too.